src/packages/packages.controller.ts

```diff
@@ -0,0 +1,198 @@
+    }
+  
+    const versions = pkg.versions.getItems();
+    if (versions.length === 0) {
+      throw new NotFoundException(`No versions found for package "${name}"`);
+    }
+  
+    const latestVersion = versions.reduce((latest, current) => {
+      if (!latest) return current;
+      
+      if (current.major > latest.major) return current;
+      if (current.major < latest.major) return latest;
+      
+      if (current.minor > latest.minor) return current;
+      if (current.minor < latest.minor) return latest;
+      
+      if (current.patch > latest.patch) return current;
+      if (current.patch < latest.patch) return latest;
+      
+      return latest;
+    });
+    const download = new Download();
+    download.package = pkg;
+    download.version = latestVersion;
+    await this.em.persistAndFlush(download);
+  
+    const buffer = Buffer.from(latestVersion.packageBlob, 'base64');
+    const fileStream = new stream.PassThrough();
+    fileStream.end(buffer);
+  
+    const fileName = `${name}-${latestVersion.versionNumber}.tar.gz`;
+    res.set({
+      'Content-Type': 'application/octet-stream',
+      'Content-Disposition': `attachment; filename="${fileName}"`,
+      'Content-Length': buffer.length,
+    });
+  
+    fileStream.pipe(res);
+  }
+  
+  @Get(':name/:version/download')
+  async downloadPackage(
+    @Param('name') name: string,
+    @Param('version') version: string,
+    @Res() res: Response,
+  ) {
+    const [major, minor, patch] = version.split('.').map(Number);
+  
+    const pkg = await this.em.findOne(Package, { name }, { populate: ['versions'] });
+    if (!pkg) {
+      throw new NotFoundException(`Package "${name}" not found`);
+    }
+  
+    const ver = pkg.versions.getItems().find(
+      v => v.major === major && v.minor === minor && v.patch === patch,
+    );
+  
+    if (!ver) {
+      throw new NotFoundException(`Version "${version}" not found for package "${name}"`);
+    }
+  
+    const download = new Download();
+    download.package = pkg;
+    download.version = ver;
+    await this.em.persistAndFlush(download);
+  
+    const buffer = Buffer.from(ver.packageBlob, 'base64');
+    const fileStream = new stream.PassThrough();
+    fileStream.end(buffer);
+  
+    const fileName = `${name}-${version}.tar.gz`; 
+    res.set({
+      'Content-Type': 'application/octet-stream',
+      'Content-Disposition': `attachment; filename="${fileName}"`,
+      'Content-Length': buffer.length, 
+    });
+  
+    fileStream.pipe(res);
+  }
+  
+  @Get(':name/:version/downloads')
+  async getDownloadsHistory(
+    @Param('name') name: string,
+    @Param('version') version: string,
+  ) {
+    const pkg = await this.em.findOne(Package, { name });
+    if (!pkg) {
+      throw new NotFoundException(`Package "${name}" not found`);
+    }
+  
+    const [major, minor, patch] = version.split('.').map(Number);
+    
+    const ver = await this.em.findOne(Version, { 
```

Comment:
If you make Package having `name` as primary key, then you don't need to fetch Package first, but you can find Version by package name. It will simplify this code (and above). Create a query to find Version by package name (trim it).